VariantShredding.md

```diff
@@ -33,140 +33,247 @@ This document focuses on the shredding semantics, Parquet representation, implic
+
+For example, a `tags` Variant may be shredded as a list of strings using the following definition:
+```
+optional group tags (VARIANT) {
+  required binary metadata;
+  optional binary variant_value;
+  optional group typed_value (LIST) { // must be optional to allow a null list
+    repeated group list {
+      required group element {
+        optional binary variant_value;
+        optional binary typed_value (STRING);
+      }
     }
-   }
   }
- }
 }
 ```
 
-| Variant Value | Top-level variant_value | b.variant_value | a.typed_value | a.variant_value | b.object.c.typed_value | b.object.c.variant_value | Notes | 
-|---------------|-------------------------|-----------------|---------------|-----------------|------------------------|--------------------------|-------|
-| {a: 123, b: {c: “hello”}} | null | null | 123 | null | hello | null | All values shredded |
-| {a: 1.23, b: {c: “123”}} | null | null | null | 1.23 | 123 | null | a is not an integer |
-| {a: 123, b: {c: null}} | null | null | null | 123 | null | null | b.object.c set to non-null to indicate VariantNull |
-| {a: 123, b: {} | null | null | null | 123 | null | null | b.object.c set to null, to indicate that c is missing |
-| {a: 123, d: 456} | {d: 456} | null | 123 | null | null | null | Extra field d is stored as variant_value |
-| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null | null | null | Not an object |
+All elements of an array must be non-null, since `array` elements cannote be missing.
+Either `typed_value` or `variant_value` (but not both) must be non-null.
+
+#### Objects (Option 1)
 
-# Parquet Layout
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-The `array` and `object` fields represent Variant array and object types, respectively.
-Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+If the value is not an object, `typed_value` must be null.
 
-An `object` field must be a group.
-Each field name of this inner group corresponds to the Variant value's object field name.
-Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of `object`, `array`, `typed_value` or `variant_value`.
+<!-- TODO: Why not keep non-shredded fields in the object itself? -->
+If the value is a partially shredded object, the `variant_value` must not contain shredded fields. If such fields are present, the object is invalid and readers must either fail or use the values from the `variant_value`.
 
-Similarly the elements of an `array` must be a group containing one or more of `object`, `array`, `typed_value` or `variant_value`.
+Each shredded field is represented as a required group that contains a `variant_value` and a `typed_value` field.
```

Comment:
Why each shredded field  should be a required group is not clear to me. If fields were allowed to be optional, that would be another way of indicating non-existence of fields.

Reply:
The primary purpose is to reduce the number of cases that implementers have to deal with. If all of the cases can be expressed with 2 optional fields rather than 2 optional fields inside an optional group, then the group should be required to simplify as much as possible.

In addition, every level in Parquet that is optional introduces another repetition/definition level. That adds up quickly with nested structures and ends up taking unnecessary space.